crates/ty_python_semantic/src/types.rs

```diff
@@ -4609,14 +4640,19 @@ impl<'db> Type<'db> {
                         )
                     })
                     .unwrap_or(instance_ty);
-                Ok(specialized)
+
+                // Return the __new__ type if available, otherwise return instance_ty
+                Ok(new_return_type.unwrap_or(specialized))
```

Comment:
It seems like if we have a `__new__` return type, but we also called `__init__`, we may need to further specialize `new_return_type` here.